File: Programmers/크레인 인형뽑기 게임.cpp
#include <string>
#include <vector>
#include <stack>
using namespace std;

int solution(vector<vector<int>> board, vector<int> moves) {
    int answer = 0;
    stack<int> s;
    for(int i=0; i<moves.size(); i++){
        int pick = moves[i] - 1;
        int h = 0;
        while(h < board.size() && board[h][pick] == 0) h++;
        if(h < board.size() && board[h][pick] != 0){
            if(!s.empty() && s.top() == board[h][pick]) {
                answer += 2;
                s.pop();
            }
            else s.push(board[h][pick]);
            board[h][pick] = 0;
        }
    }
    return answer;
}

File: Programmers/크레인 인형뽑기 게임.java
import java.util.Stack;

class Solution {
    public int solution(int[][] board, int[] moves) {
        int answer = 0;
        int n = board.length;
        Stack<Integer> stack = new Stack<>();
        for (int i = 0; i < moves.length; i++) {
            int x = moves[i] - 1;
            for (int y = 0; y < n; y++) {
                if (board[y][x] != 0) {
                    int doll = board[y][x];
                    board[y][x] = 0;
                    if (!stack.empty() && stack.peek() == doll) {
                        stack.pop();
                        answer += 2;
                    } else {
                        stack.push(doll);
                    }
                    break;
                }
            }
        }
        return answer;
    }
}